public/app/containers/Explore/PromQueryField.tsx

```diff
@@ -0,0 +1,282 @@
+      context = 'context-aggregation';
+      const metric = metricNode.textContent;
+      const labelKeys = this.state.labelKeys[metric];
+      if (labelKeys) {
+        suggestions.push({ label: 'Labels', items: labelKeys.map(wrapText) });
+      } else {
+        refresher = this.fetchMetricLabels(metric);
+      }
+    } else if (
+      (this.state.metrics && ((prefix && !wrapperClasses.contains('token')) || text.match(/[+\-*/^%]/))) ||
+      wrapperClasses.contains('context-function')
+    ) {
+      // Need prefix for metrics
+      context = 'context-metrics';
+      suggestions.push({
+        label: 'Metrics',
+        items: this.state.metrics.map(wrapText),
+      });
+    }
+
+    console.log('handleTypeahead', selection.anchorNode, wrapperClasses, text, offset, prefix, context);
+
+    return {
+      context,
+      prefix,
+      refresher,
+      suggestions,
+    };
+  };
+
+  async fetchLabelValues(key) {
+    const url = `/api/v1/label/${key}/values`;
+    try {
+      const res = await this.request(url);
+      const body = await (res.data || res.json());
+      const pairs = this.state.labelValues[EMPTY_METRIC];
+      const values = {
+        ...pairs,
+        [key]: body.data,
+      };
+      // const labelKeys = {
+      //   ...this.state.labelKeys,
+      //   [EMPTY_METRIC]: keys,
+      // };
+      const labelValues = {
+        ...this.state.labelValues,
+        [EMPTY_METRIC]: values,
+      };
+      this.setState({ labelValues });
+    } catch (e) {
+      if (this.props.onRequestError) {
+        this.props.onRequestError(e);
+      } else {
+        console.error(e);
+      }
+    }
+  }
+
+  async fetchMetricLabels(name) {
+    const url = `/api/v1/series?match[]=${name}`;
+    try {
+      const res = await this.request(url);
+      const body = await (res.data || res.json());
+      const { keys, values } = processLabels(body.data);
+      const labelKeys = {
+        ...this.state.labelKeys,
+        [name]: keys,
+      };
+      const labelValues = {
+        ...this.state.labelValues,
+        [name]: values,
+      };
+      this.setState({ labelKeys, labelValues });
+    } catch (e) {
+      if (this.props.onRequestError) {
+        this.props.onRequestError(e);
+      } else {
+        console.error(e);
+      }
+    }
+  }
+
+  async fetchMetricNames() {
+    const url = '/api/v1/label/__name__/values';
+    try {
+      const res = await this.request(url);
+      const body = await (res.data || res.json());
+      this.setState({ metrics: body.data }, this.onMetricsReceived);
+    } catch (error) {
+      if (this.props.onRequestError) {
+        this.props.onRequestError(error);
+      } else {
+        console.error(error);
+      }
+    }
+  }
+
+  render() {
+    return (
+      <TypeaheadField
+        additionalPlugins={this.plugins}
+        cleanText={cleanText}
+        initialValue={this.props.initialQuery}
+        onTypeahead={this.handleTypeahead}
```

Comment:
I think we should align on the airbnb style guide of naming event / callback functions "onTypeahead" , onSomethingChanged etc, instead of handle prefix.